src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -742,5 +764,100 @@ private void UpdateCacheExpirationTime(KeyValueWatcher changeWatcher, bool succe
+
+        private async Task ExecuteWithFailOverPolicyAsync(Func<ConfigurationClient, Task> funcToExecute, CancellationToken cancellationToken = default)
+        {
+            using IEnumerator<ConfigurationClient> clientEnumerator = _configClientProvider.GetClients().GetEnumerator();
+
+            // Guaranteed to return at least 1 element
+            Debug.Assert(clientEnumerator.MoveNext());
+
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                try
+                {
+                    await funcToExecute(clientEnumerator.Current).ConfigureAwait(false);
+
+                    _configClientProvider.UpdateClientStatus(clientEnumerator.Current, successful: true);
+
+                    break;
+                }
+                catch (AggregateException ae)
+                {
+                    if (!ShouldFailoverForException(ae))
+                    {
+                        throw;
+                    }
+
+                    _configClientProvider.UpdateClientStatus(clientEnumerator.Current, successful: false);
+
+                    if (!clientEnumerator.MoveNext())
+                    {
+                        throw;
+                    }
+                }
+                catch (RequestFailedException e)
+                {
+                    if (!ShouldFailoverForException(e))
+                    {
+                        throw;
+                    }
+
+                    _configClientProvider.UpdateClientStatus(clientEnumerator.Current, successful: false);
+
+                    if (!clientEnumerator.MoveNext())
+                    {
+                        throw;
+                    }
+                }
+            }
+        }
+
+        private void UpdateCacheExpirationForWatchers(bool success)
+        {
+            // Update the cache expiration time for all refresh registered settings and feature flags
+            foreach (KeyValueWatcher changeWatcher in _options.ChangeWatchers.Concat(_options.MultiKeyWatchers))
+            {
+                UpdateCacheExpirationTime(changeWatcher, success);
+            }
+
+        }
+
+        private bool ShouldFailoverForException(Exception ex)
+        {
+            if (ex is AggregateException aggregateException)
+            {
+                IReadOnlyCollection<Exception> innerExceptions = aggregateException.InnerExceptions;
+
+                if (innerExceptions.Any() && innerExceptions.All(ex => ex is RequestFailedException))
+                {
+                    return ShouldFailOverForRequestFailedException(innerExceptions.Last() as RequestFailedException);
+                }
+            }
+            else if (ex is RequestFailedException rfe)
+            {
+                return ShouldFailOverForRequestFailedException(rfe);
+            }
+
+            return false;
+        }
+
+        private bool ShouldFailOverForRequestFailedException(RequestFailedException rfe)

```

Comment:
I'd recommend just `ShouldFailOver` the `ForRequestFailedException` is encapsulated by the parameter.

Reply:
Makes sense! Done!